htmresearch/algorithms/column_pooler.py

```diff
@@ -207,43 +264,53 @@ def _computeLearningMode(self, feedforwardInput, lateralInputs,
 
     @param  feedforwardGrowthCandidates (sequence or None)
             Sorted indices of feedforward input bits that the active cells may
-            grow new synapses to
+            grow new synapses to.  This is assumed to be the predicted active
+            cells of the input layer.
     """
-
     prevActiveCells = self.activeCells
 
-    # If there are no previously active cells, select random subset of cells.
-    # Else we maintain previous activity.
-    if len(self.activeCells) == 0:
+    # If there are not enough previously active cells, then we are no longer on
+    # a familiar object.  Either our representation decayed due to the passage
+    # of time (i.e. we moved somewhere else) or we were mistaken.  Either way,
+    # create a new SDR and learn on it.
+    # This case is the only way different object representations are created.
+    if len(self.activeCells) <= self.sdrSize*(1 - self.learningTolerance):
       self.activeCells = _sampleRange(self._random,
                                       0, self.numberOfCells(),
                                       step=1, k=self.sdrSize)
       self.activeCells.sort()
 
+    # If we have a union of cells active, don't learn.  This primarily affects
+    # online learning.
+    if len(self.activeCells) >= self.sdrSize*(1+self.learningTolerance):
+      return
+
+    # Finally, now that we have decided which cells we should be learning on, do
+    # the actual learning.
     if len(feedforwardInput) > 0:
-      # Proximal learning
       self._learn(self.proximalPermanences, self._random,
                   self.activeCells, feedforwardInput,
                   feedforwardGrowthCandidates, self.sampleSizeProximal,
                   self.initialProximalPermanence, self.synPermProximalInc,
                   self.synPermProximalDec, self.connectedPermanenceProximal)
 
-      # Internal distal learning
-      if len(prevActiveCells) > 0:
-        self._learn(self.internalDistalPermanences, self._random,
-                    self.activeCells, prevActiveCells, prevActiveCells,
-                    self.sampleSizeDistal, self.initialDistalPermanence,
-                    self.synPermDistalInc, self.synPermDistalDec,
-                    self.connectedPermanenceDistal)
-
       # External distal learning
+      # We should do this no matter what, since other columns might still be
+      # learning useful things even if we're not.
```

Comment:
Sorry, this has gotten bounced back and forth several times.  I corrected the previous version to learn constantly, but then it had to be changed back when I merged the two learning methods, and I forgot to update the comment.